pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,315 @@
+	}
+	ret, found := resolved[provider]
+	if !found {
+		var li []string
+		for p := range resolved {
+			li = append(li, p)
+		}
+		errs = multierror.Append(fmt.Errorf("extension provider %s not found, available providers are %v", provider, li))
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	authzLog.Debugf("resolved provider %s to config: %v", provider, spew.Sdump(ret))
+	return ret, nil
+}
+
+func buildExtAuthzHTTP(config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationHttpProvider) (*builtExtAuthz, error) {
+	var errs error
+	port, err := parsePort(config.Port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	hostname, cluster, err := parseService(config.Service, port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	status, err := parseStatusOnError(config.StatusOnError)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if config.PathPrefix != "" {
+		if _, err := url.Parse(config.PathPrefix); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("invalid pathPrefix %q: %v", config.PathPrefix, err))
+		}
+		if !strings.HasPrefix(config.PathPrefix, "/") {
+			errs = multierror.Append(errs, fmt.Errorf("pathPrefix must begin with `/`, found: %q", config.PathPrefix))
+		}
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateHTTPConfig(hostname, cluster, status, config), nil
+}
+
+func buildExtAuthzGRPC(config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationGrpcProvider) (*builtExtAuthz, error) {
+	var errs error
+	port, err := parsePort(config.Port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	_, cluster, err := parseService(config.Service, port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	status, err := parseStatusOnError(config.StatusOnError)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateGRPCConfig(cluster, config.FailOpen, status), nil
+}
+
+func parseService(service string, port int) (hostname string, cluster string, err error) {
+	if service == "" {
+		return "", "", fmt.Errorf("service must not be empty")
+	}
+	parts := strings.Split(service, "/")
+	if len(parts) > 2 {
+		return "", "", fmt.Errorf("service not in format [<namespace>/]<service>, found: %s", service)
+	}
+	var name, namespace string
+	if len(parts) == 2 {
+		namespace, name = parts[0], parts[1]
+	} else {
+		// TODO(yangminzhu): Fix to use the namespace of MeshConfig.
+		namespace, name = "istio-system", parts[0]
+	}
+
+	// TODO(yangminzhu): This is temporary and still under development.
+	// The remaining work is to properly get the corresponding cluster name for the k8s service and ServiceEntry.
+	hostname = fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace)
+	cluster = model.BuildSubsetKey(model.TrafficDirectionOutbound, "", host.Name(hostname), port)
+	return hostname, cluster, nil
+}
+
+func parsePort(port uint32) (int, error) {
```

Comment:
Are there some model package util function for this kinda definition? I feel very likely

Reply:
there is one in validation pkg but it caused some cyclic import issue.